src/logger/logger.cpp

```diff
@@ -0,0 +1,84 @@
+#include "logger.hh"
+
+#include <cstdarg>
+#include <iomanip>
+#include <iostream>
+#include <filesystem>
+#include <string>
+#include <thread>
+
+ZarrLogLevel Logger::current_level_ = ZarrLogLevel_Info;
+std::mutex Logger::log_mutex_{};
+
+void
+Logger::set_log_level(ZarrLogLevel level)
+{
+    current_level_ = level;
+}
+
+ZarrLogLevel
+Logger::get_log_level()
+{
+    return current_level_;
+}
+
+std::string
+Logger::log(ZarrLogLevel level,
+            const char* file,
+            int line,
+            const char* func,
+            const char* format,
+            ...)
+{
+    std::scoped_lock lock(log_mutex_);
+    if (current_level_ == ZarrLogLevel_None || level < current_level_) {
+        return {}; // Suppress logs
+    }
+
+    va_list args;
+    va_start(args, format);
```

Comment:
I mentioned this in the comprehensive PR review, but a modern approach to handling an arbitrary number of arguments is using variadic templates.

| **Aspect**               | **`va_list` (C)**                                    | **Variadic Templates (C++)**                       |
|--------------------------|-----------------------------------------------------|---------------------------------------------------|
| **Type Safety**           | None, relies on the programmer to manage types      | Full type safety, checked by the compiler         |
| **Runtime vs Compile-time** | Arguments processed at runtime                     | Arguments processed at compile time               |
| **Ease of Use**           | Requires manual handling with macros                | Recursion or fold expressions make usage simpler  |
| **Flexibility**           | Limited to basic types and needs type casting       | Supports any type (built-in and user-defined)     |
| **Error Handling**        | Type errors lead to runtime crashes or UB           | Errors caught at compile-time                     |
| **Performance**           | Slower (runtime argument parsing)                   | Faster (resolved at compile-time)                 |
| **Common Use Cases**      | C-style `printf`, legacy C functions                | Generic functions or classes in modern C++        |